File: JavaScript/mean_median.js
var n = prompt("enter the size of an array required : ");
n = parseInt(n);
var arr = new Array();
//taking input for array
for (let i = 0; i < n; i++) {
	arr[i] = prompt("element " + (i + 1) + ":");
	arr[i] = parseInt(arr[i]);
}

//to find mean of a given array
var sum = 0;
for (let i = 0; i < n; i++) {
	sum = sum + arr[i]
}
console.log("Mean of a given array is : ", sum / n);
//to find median of a given array

//sort the given array
arr1 = arr.sort();
var half = Math.floor(arr1.length / 2);
if (n % 2 == 0) {
	console.log("Median of a given array is : ", (arr1[half] + arr1[half - 1]) / 2.0);
} else
	console.log("Median of a given array is : ", arr1[half]);
  
  
  /*
  problem: to find mean and median of a given array
  sample I/O:
  enter the size of an array required : 7
  element 1:10
  element 2:23
  element 3:2
  element 4:21
  element 5:23
  element 6:38
  element 7:38
  Mean of a given array is : 22.142857142857
  Meadian of a given array is : 23
  
  space complexity: O(log(n))
  time complexity: O(nlog(n))
  */

File: C-Plus-Plus/mean_median.cpp
/* 
 *
 * Topic: Mean and Median calculation
 * Language: C++
 * 
 */

#include <bits/stdc++.h>

using namespace std;

/*
 *
 * Mean of an array = (sum of all elements) / (number of elements)
 * 
 * Time complexity: O(n)
 * 
 */
int find_mean(int arr[], int n)
{
    int sum = 0;
    for (auto i = 0; i < n; i++)
        sum += arr[i];
    return sum / n;
}

/*
 *
 * Median of a sorted array of size n is defined as the middle element
 * when n is odd and average of middle two elements when n is even.
 * 
 * Time Complexity: O(nlogn)
 * 
 */
int find_median(int arr[], int n)
{
    if (n == 0)
        return -1;
    sort(arr, arr + n);
    return n % 2 == 0 ? (arr[n / 2 - 1] + arr[n / 2]) / 2 : arr[n / 2];
}

int main()
{
    // Size of the array
    int n;
    cout << "Enter the size of the array: ";
    cin >> n;
    int arr[n];
    cout << "\nEnter the array elements: ";
    for (auto i = 0; i < n; i++)
        cin >> arr[i];

    int mean = find_mean(arr, n);
    int median = find_median(arr, n);

    cout << "\nMean of the array is: " << mean;
    cout << "\nMedian of the array is: " << median;
    return 0;
}

/*
 *
 * Input:
 * Enter the size of the array: 6
 * Enter the array elements: 22 18 11 14 20 11
 * 
 * Output:
 * Mean of the array is: 16
 * Median of the array is: 16
 * 
 */

